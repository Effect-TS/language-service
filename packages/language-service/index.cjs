// eslint-disable-next-line @typescript-eslint/consistent-type-imports, no-undef
module.exports = function init(modules) {
  function create(info) {
    // start with regular language service, then load custom once async module has loaded
    let languageService = info.languageService
    import("./build/cjs/index.js").then(v => languageService = v.default(modules).create(info))

    // create the proxy
    const proxy = Object.create(null)
    for (const k of Object.keys(info.languageService)) {
      proxy[k] = (...args) => languageService[k].apply(languageService, args)
    }

    return proxy
  }

  return { create }
}
