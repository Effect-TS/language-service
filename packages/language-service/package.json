{
  "name": "@effect/language-service",
  "version": "0.0.0",
  "type": "module",
  "repository": "https://github.com/Effect-TS/language-service.git",
  "author": "Mattia Manzati",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "scripts": {
    "clean": "yarn org:rimraf build tsbuildinfo dist",
    "lint": "yarn org:eslint . --ext .ts,.tsx",
    "autofix": "yarn lint --fix"
  },
  "main": "./index.cjs",
  "exports": {
    ".": {
      "import": "./build/cjs/index.js"
    },
    "./*": {
      "import": "./build/cjs/*.js"
    },
    "./test/*": {
      "import": "./build/test/*.js"
    }
  },
  "dependencies": {
    "@effect/core": "^0.0.15",
    "@tsplus/runtime": "^0.1.10",
    "@tsplus/stdlib": "^0.2.8"
  },
  "config": {
    "side": [],
    "modules": []
  }
}
