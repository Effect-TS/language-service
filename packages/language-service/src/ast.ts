import { Tag } from "@tsplus/stdlib/service/Tag"
import type ts from "typescript/lib/tsserverlibrary"

declare module "typescript/lib/tsserverlibrary" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any
  export type ChangeNodeOptions = any

  export namespace textChanges {
    export class ChangeTracker {
      static with(context: ts.TextChangesContext, cb: (tracker: ChangeTracker) => void): ts.FileTextChanges[]
      replaceNode(sourceFile: ts.SourceFile, oldNode: ts.Node, newNode: ts.Node, options?: ts.ChangeNodeOptions): void
    }
    export function applyChanges(text: string, changes: readonly ts.TextChange[]): string
  }

  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFileLike,
    startNode: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFile,
    startNode?: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined

  export function isMemberName(node: ts.Node): node is ts.MemberName
  export function isKeyword(token: ts.SyntaxKind): token is ts.KeywordSyntaxKind
}

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export type TypeScriptApi = typeof import("typescript/lib/tsserverlibrary")
export const TypeScriptApi = Tag<TypeScriptApi>()
export const LanguageServiceApi = Tag<ts.LanguageService>()
export const ChangeTrackerApi = Tag<ts.textChanges.ChangeTracker>()

export class NoTypeScriptProgramError {
  readonly _tag = "NoTypeScriptProgramError"
}

export function getProgram() {
  return Effect.serviceWithEffect(
    LanguageServiceApi,
    languageService => Effect.sync(() => languageService.getProgram())
  )
    .filterOrFail((program): program is ts.Program => !!program, () => new NoTypeScriptProgramError())
}

export class NoSuchSourceFile {
  readonly _tag = "NoSuchSourceFile"
  constructor(
    readonly fileName: string
  ) {}
}

export function getSourceFile(fileName: string) {
  return getProgram().map(program => program.getSourceFile(fileName))
    .filterOrFail((sourceFile): sourceFile is ts.SourceFile => !!sourceFile, () => new NoSuchSourceFile(fileName))
}

export function hasModifier(node: ts.Declaration, kind: ts.ModifierFlags) {
  return Effect.serviceWith(TypeScriptApi, ts => !!(ts.getCombinedModifierFlags(node) & kind))
}

function getChildNodesContainingRange(
  parent: ts.Node,
  textRange: ts.TextRange
) {
  return Do($ => {
    const ts = $(Effect.service(TypeScriptApi))
    let result: Chunk<ts.Node> = Chunk.empty()

    ts.forEachChild(parent, node => {
      if (node.end < textRange.pos) {
        return
      } else if (node.pos > textRange.end) {
        return
      } else {
        result = result.append(node)
      }
    })

    return result
  })
}

/**
 * Gets the closest node that contains given TextRange
 */
export function getNodesContainingRange(
  sourceFile: ts.SourceFile,
  textRange: ts.TextRange
) {
  return Do(($) => {
    let result: Chunk<ts.Node> = Chunk.empty()

    $(
      Effect.iterate([sourceFile] as ts.Node[], (_) => _.length > 0)(toProcess => {
        const node = toProcess.shift()!
        return getChildNodesContainingRange(node, textRange).flatMap(_ =>
          Effect.sync(() => {
            result = result.concat(_)
            return toProcess.concat(..._.toArray)
          })
        )
      })
    )

    return Chunk.from(result.reverse)
  })
}

/**
 * Ensures value is a text range
 */
export function toTextRange(positionOrRange: number | ts.TextRange): ts.TextRange {
  return typeof positionOrRange === "number" ? { end: positionOrRange, pos: positionOrRange } : positionOrRange
}

export function getHumanReadableName(sourceFile: ts.SourceFile, node: ts.Node) {
  const text = node.getText(sourceFile)
  return text.length > 10 ? text.substring(0, 10) + "..." : text
}

export function collectAll<R, E>(rootNode: ts.Node, test: (node: ts.Node) => Effect<R, E, boolean>) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const env = $(Effect.environment<R>())
    let result = Chunk.empty<ts.Node>()

    function visitor(node: ts.Node) {
      if (test(node).provideEnvironment(env).unsafeRunSync()) result = result.append(node)
      ts.visitEachChild(node, visitor, ts.nullTransformationContext)

      return node
    }

    ts.visitNode(rootNode, visitor)

    return result
  })
}

export function getRelevantTokens(
  position: number,
  sourceFile: ts.SourceFile
) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const previousToken = ts.findPrecedingToken(position, sourceFile)
    if (
      previousToken && position <= previousToken.end &&
      (ts.isMemberName(previousToken) || ts.isKeyword(previousToken.kind))
    ) {
      const contextToken = ts.findPrecedingToken(previousToken.getFullStart(), sourceFile, /*startNode*/ undefined)! // TODO: GH#18217
      return { contextToken, previousToken }
    }
    return { contextToken: Maybe.fromNullable(previousToken), previousToken: Maybe.fromNullable(previousToken) }
  })
}
