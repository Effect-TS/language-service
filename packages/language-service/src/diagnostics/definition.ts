import type ts from "typescript/lib/tsserverlibrary"

export interface DiagnosticDefinition {
  code: number
  apply: <E>(
    sourceFile: ts.SourceFile
  ) => Effect<
    typeof ts | ts.LanguageService,
    E,
    Chunk<{
      node: ts.Node
      category: ts.DiagnosticCategory
      messageText: string
    }>
  >
}

export function createDiagnostic(definition: DiagnosticDefinition) {
  return definition
}
