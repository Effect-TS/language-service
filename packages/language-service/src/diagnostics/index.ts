import removeCurryArrow from "@effect/language-service/diagnostics/removeCurryArrow"

export default { removeCurryArrow }
