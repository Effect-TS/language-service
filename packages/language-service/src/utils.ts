import * as AST from "@effect/language-service/ast"
import type ts from "typescript/lib/tsserverlibrary"

export function isPipeCall(node: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    if (!ts.isCallExpression(node)) return false
    const expression = node.expression
    if (!ts.isIdentifier(expression)) return false
    if (expression.getText(node.getSourceFile()) !== "pipe") return false
    return true
  })
}

export function asPipeableCallExpression(node: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    // ensure the node is a call expression
    if (!ts.isCallExpression(node)) return Maybe.none
    // with just 1 arg
    if (node.arguments.length !== 1) return Maybe.none
    const arg = node.arguments[0]!
    // ideally T.map(n => n * 2) could be piped to pipe(n => n * 2, T.map)
    // but does not make any sense.
    if (ts.isArrowFunction(arg)) return Maybe.none
    // same goes for identifiers, string literal or numbers
    if (ts.isStringLiteral(arg) || ts.isNumericLiteral(arg) || ts.isIdentifier(arg)) return Maybe.none
    return Maybe.some([node.expression, arg] as const)
  })
}

export function asPipeArguments(initialNode: ts.Node) {
  return Do($ => {
    let result = Chunk.empty<ts.Expression>()
    $(
      Effect.iterate(Maybe.some(initialNode), node => node.isSome())(
        maybeNode =>
          Do($ => {
            if (maybeNode.isNone()) return Maybe.none

            const node = maybeNode.value
            const maybePipeable = $(asPipeableCallExpression(node))
            if (maybePipeable.isNone()) {
              result = result.append(node as ts.Expression)
              return Maybe.none
            }
            const [exp, arg] = maybePipeable.value
            result = result.append(exp)

            return Maybe.some(arg)
          })
      )
    )
    return result.reverse
  })
}

export function isPipeableCallExpression(node: ts.Node) {
  return asPipeableCallExpression(node).map(_ => _.isSome())
}

export function isCurryArrow(arrow: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    if (!ts.isArrowFunction(arrow)) return false
    if (arrow.parameters.length !== 1) return false
    const parameter = arrow.parameters[0]!
    const parameterName = parameter.name
    if (!ts.isIdentifier(parameterName)) return false
    const body = arrow.body
    if (!ts.isCallExpression(body)) return false
    const args = body.arguments
    if (args.length !== 1) return false
    const identifier = args[0]!
    if (!ts.isIdentifier(identifier)) return false
    return identifier.text === parameterName.text
  })
}

export function transformAsyncAwaitToEffectGen(
  node: ts.FunctionDeclaration | ts.ArrowFunction | ts.FunctionExpression,
  onAwait: (expression: ts.Expression) => ts.Expression
) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))

    function visitor(_: ts.Node): ts.Node {
      if (ts.isAwaitExpression(_)) {
        const expression = ts.visitEachChild(_.expression, visitor, ts.nullTransformationContext)

        return ts.factory.createYieldExpression(
          ts.factory.createToken(ts.SyntaxKind.AsteriskToken),
          ts.factory.createCallExpression(ts.factory.createIdentifier("$"), undefined, [onAwait(expression)])
        )
      }
      return ts.visitEachChild(_, visitor, ts.nullTransformationContext)
    }
    const generatorBody = visitor(node.body!)

    const generator = ts.factory.createFunctionExpression(
      [],
      ts.factory.createToken(ts.SyntaxKind.AsteriskToken),
      undefined,
      [],
      [ts.factory.createParameterDeclaration(undefined, undefined, "$")],
      undefined,
      generatorBody as any // NOTE(mattia): intended, to use same routine for both ConciseBody and Body
    )

    const effectGenCallExp = ts.factory.createCallExpression(
      ts.factory.createPropertyAccessExpression(
        ts.factory.createIdentifier("Effect"),
        "gen"
      ),
      undefined,
      [generator as any]
    )

    let currentFlags = ts.getCombinedModifierFlags(node)
    currentFlags &= ~ts.ModifierFlags.Async
    const newModifiers = ts.factory.createModifiersFromModifierFlags(currentFlags)

    if (ts.isArrowFunction(node)) {
      return ts.factory.createArrowFunction(
        newModifiers,
        node.typeParameters,
        node.parameters,
        undefined,
        node.equalsGreaterThanToken,
        effectGenCallExp
      )
    }

    const newBody = ts.factory.createBlock([
      ts.factory.createReturnStatement(effectGenCallExp)
    ])

    if (ts.isFunctionDeclaration(node)) {
      return ts.factory.createFunctionDeclaration(
        newModifiers,
        node.asteriskToken,
        node.name,
        node.typeParameters,
        node.parameters,
        undefined,
        newBody
      )
    }
    return ts.factory.createFunctionExpression(
      newModifiers,
      node.asteriskToken,
      node.name,
      node.typeParameters,
      node.parameters,
      undefined,
      newBody
    )
  })
}
