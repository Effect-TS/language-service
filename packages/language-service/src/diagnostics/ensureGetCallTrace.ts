import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import { createDiagnostic } from "@effect/language-service/diagnostics/definition"

export default createDiagnostic({
  code: 1003,
  apply: (sourceFile) =>
    Do($ => {
      const ts = $(T.service(AST.TypeScriptApi))
      const program = $(T.service(AST.TypeScriptProgram))
      const typeChecker = program.getTypeChecker()

      function isEffectType(type: ts.Type) {
        const symbol = type.getSymbol()
        if (!symbol) return false
        return symbol.getDocumentationComment(typeChecker).some(i =>
          i.text.indexOf("Effects model resourceful interaction") > -1
        )
      }

      function isPipeableCombinator(type: ts.Type) {
        const signatures = type.getCallSignatures()
        if (signatures.length === 0) return false
        return signatures.some(isPipeableCallSignatureReturningEffect)
      }

      function isPipeableCallSignatureReturningEffect(signature: ts.Signature) {
        if (signature.getParameters().length !== 1) return false
        return isEffectType(signature.getReturnType())
      }

      function shouldBeTraced(signature: ts.Signature) {
        const returnType = signature.getReturnType()
        return isEffectType(returnType) || isPipeableCombinator(returnType)
      }

      function isGetCallTraceCallExpression(node: ts.Node): node is ts.CallExpression {
        if (!ts.isCallExpression(node)) return false
        const expression = node.expression
        if (!ts.isIdentifier(expression)) return false
        return expression.text === "getCallTrace"
      }

      const definitions = AST.collectAll(ts)(sourceFile, ts.isArrowFunction).concat(
        AST.collectAll(ts)(sourceFile, ts.isFunctionDeclaration)
      )

      return definitions.filter(node => !!node.body).filter(node =>
        typeChecker.getTypeAtLocation(node).getCallSignatures().some(shouldBeTraced)
      ).filter(node => {
        const body = node.body!
        if (!ts.isBlock(body)) return true
        const firstStatement = body.statements[0]
        if (!firstStatement) return true
        return AST.collectAll(ts)(firstStatement, isGetCallTraceCallExpression).isEmpty
      }).map((node) => ({
        node,
        category: ts.DiagnosticCategory.Warning,
        messageText: `This function should be traced. getCallTrace() should be the first line of the body.`
      }))
    })
})
