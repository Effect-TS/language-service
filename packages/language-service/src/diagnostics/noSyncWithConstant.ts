import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import type { DiagnosticDefinitionMessage } from "@effect/language-service/diagnostics/definition"
import { createDiagnostic } from "@effect/language-service/diagnostics/definition"
import {
  findModuleImportIdentifierName,
  isCombinatorCall,
  isLiteralConstantValue
} from "@effect/language-service/utils"

export default createDiagnostic({
  code: 1002,
  apply: (sourceFile) =>
    Do($ => {
      const ts = $(T.service(AST.TypeScriptApi))

      const effectIdentifier = findModuleImportIdentifierName(ts)(sourceFile, "@effect/core/io/Effect").getOrElse(
        "Effect"
      )

      const methodsMap = {
        sync: "succeed",
        failSync: "fail",
        dieSync: "die"
      }

      let result: Chunk<DiagnosticDefinitionMessage> = Chunk.empty()

      for (const methodName of Object.keys(methodsMap)) {
        const suggestedMethodName: string = methodsMap[methodName]!

        result = result.concat(
          AST.collectAll(ts)(sourceFile, isCombinatorCall(ts)(effectIdentifier, methodName)).filter(node => {
            if (node.arguments.length !== 1) return false
            const arg = node.arguments[0]!
            if (!ts.isArrowFunction(arg)) return false
            if (!isLiteralConstantValue(ts)(arg.body)) return false
            return true
          }).map(node => ({
            node,
            category: ts.DiagnosticCategory.Warning,
            messageText: `Value is constant, instead of using ${methodName} you could use ${suggestedMethodName}.`
          }))
        )
      }

      return result
    })
})
