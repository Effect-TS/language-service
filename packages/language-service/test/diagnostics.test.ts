import * as AST from "@effect/language-service/ast"
import type { DiagnosticDefinition } from "@effect/language-service/diagnostics/definition"
import diagnostics from "@effect/language-service/diagnostics/index"
import { createMockLanguageServiceHost } from "@effect/language-service/test/utils"
import * as fs from "fs"
import ts from "typescript/lib/tsserverlibrary"

export function testDiagnosticOnExample(diagnostic: DiagnosticDefinition, fileName: string) {
  const sourceText = fs.readFileSync(require.resolve(__dirname + "/../../examples/diagnostics/" + fileName))
    .toString("utf8")

  // create the language service
  const languageServiceHost = createMockLanguageServiceHost(fileName, sourceText)
  const languageService = ts.createLanguageService(languageServiceHost, undefined, ts.LanguageServiceMode.Semantic)
  const sourceFile = languageService.getProgram()?.getSourceFile(fileName)
  if (!sourceFile) throw new Error("No source file " + fileName + " in VFS")

  // ensure there are no errors in TS file
  const diagnostics = languageService.getCompilerOptionsDiagnostics()
    .concat(languageService.getSyntacticDiagnostics(fileName))
    .concat(languageService.getSemanticDiagnostics(fileName)).map(diagnostic => {
      const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n")
      if (diagnostic.file) {
        const { character, line } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
        return `  Error ${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`
      } else {
        return `  Error: ${message}`
      }
    })
  expect(diagnostics).toEqual([])

  // check and assert the diagnostic runs
  const resultMessages = diagnostic
    .apply(sourceFile)
    .provideService(AST.TypeScriptApi, ts)
    .provideService(AST.LanguageServiceApi, languageService)
    .unsafeRunSync()

  // create human readable messages
  const humanMessages = resultMessages.map(error => {
    const start = ts.getLineAndCharacterOfPosition(sourceFile, error.node.pos)
    const end = ts.getLineAndCharacterOfPosition(sourceFile, error.node.end)
    const errorSourceCode = sourceText.substring(error.node.pos, error.node.end)

    return errorSourceCode + "\n" +
      `${start.line + 1}:${start.character} - ${end.line + 1}:${end.character} | ${error.messageText}`
  }).toArray.join("\n\n")
  expect(humanMessages).toMatchSnapshot()
}

function testFiles(name: string, refactor: DiagnosticDefinition, fileNames: string[]) {
  for (const fileName of fileNames) {
    describe(fileName, () => {
      it(fileName, () => {
        testDiagnosticOnExample(refactor, fileName)
      })
    })
  }
}

Object.keys(diagnostics).map(diagnosticName =>
  testFiles(diagnosticName, diagnostics[diagnosticName]!, [diagnosticName + ".ts"])
)
