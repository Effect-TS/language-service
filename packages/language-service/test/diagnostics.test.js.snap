// Vitest Snapshot v1

exports[`removeCurryArrow.ts > removeCurryArrow.ts 1`] = `
"(_) => T.log(_)
6:8 - 6:23 | This arrow function may be not needed."
`;
